File: tests/reference/r/validate_pgv_ier.R
suppressPackageStartupMessages({
  library(readxl)
  library(jsonlite)
})

dataset <- "tests/golden-data/correlation/PGV_IER_384.xlsx"
output  <- "tests/reference/r/PGV_IER_384_expected.json"

pgv_items <- paste0("PGV", 1:12)
ier_items <- paste0("IER", 1:12)

df <- read_excel(dataset, sheet = 1)

required <- c(pgv_items, ier_items)
missing_columns <- setdiff(required, names(df))

if (length(missing_columns) > 0) {
  stop(paste("Columnas faltantes:", paste(missing_columns, collapse = ", ")))
}

pgv_total <- rowSums(df[, pgv_items], na.rm = FALSE)
ier_total <- rowSums(df[, ier_items], na.rm = FALSE)

valid <- complete.cases(pgv_total, ier_total)
pgv_total <- pgv_total[valid]
ier_total <- ier_total[valid]

pearson  <- cor.test(pgv_total, ier_total, method = "pearson")
spearman <- cor.test(pgv_total, ier_total, method = "spearman", exact = FALSE)
kendall  <- cor.test(pgv_total, ier_total, method = "kendall", exact = FALSE)

baremo <- function(x) {
  categoria <- cut(
    x,
    breaks = c(-Inf, 28, 44, Inf),
    labels = c("Bajo", "Medio", "Alto"),
    include.lowest = TRUE,
    right = TRUE
  )

  if (any(is.na(categoria))) {
    stop("Existen participantes sin clasificar.")
  }

  frecuencias <- table(factor(
    categoria,
    levels = c("Bajo", "Medio", "Alto")
  ))

  resultado <- as.list(as.integer(frecuencias))
  names(resultado) <- names(frecuencias)

  if (sum(unlist(resultado)) != length(x)) {
    stop("Las frecuencias no suman N.")
  }

  resultado
}

pgv_baremo <- baremo(pgv_total)
ier_baremo <- baremo(ier_total)

stopifnot(
  length(pgv_total) == 384,
  abs(unname(spearman$estimate) - 0.7867272) < 1e-6,
  identical(unlist(pgv_baremo), c(Bajo = 115L, Medio = 143L, Alto = 126L)),
  identical(unlist(ier_baremo), c(Bajo = 79L, Medio = 218L, Alto = 87L))
)

resultado <- list(
  dataset = dataset,
  n = length(pgv_total),
  pearson = list(
    r = unname(pearson$estimate),
    p_value = pearson$p.value
  ),
  spearman = list(
    rho = unname(spearman$estimate),
    p_value = spearman$p.value
  ),
  kendall = list(
    tau_b = unname(kendall$estimate),
    p_value = kendall$p.value
  ),
  baremos_teoricos = list(
    criterio_total = list(
      Bajo = "12–28",
      Medio = "29–44",
      Alto = "45–60"
    ),
    PGV = pgv_baremo,
    IER = ier_baremo
  ),
  status = "PASS"
)

write_json(
  resultado,
  output,
  pretty = TRUE,
  auto_unbox = TRUE,
  digits = 16
)

cat(toJSON(
  resultado,
  pretty = TRUE,
  auto_unbox = TRUE,
  digits = 16
), "\n")

File: tests/reference/python/validate_pgv_ier.py
from __future__ import annotations

import hashlib
import json
import math
from pathlib import Path

import pandas as pd
from scipy import stats


ROOT = Path(__file__).resolve().parents[3]
DATASET = ROOT / "tests/golden-data/correlation/PGV_IER_384.xlsx"
OUTPUT = ROOT / "tests/reference/python/PGV_IER_384_expected.json"

PGV_ITEMS = [f"PGV{i}" for i in range(1, 13)]
IER_ITEMS = [f"IER{i}" for i in range(1, 13)]

EXPECTED = {
    "n": 384,
    "spearman_rho": 0.7867272,
    "pgv_baremo": {"Bajo": 115, "Medio": 143, "Alto": 126},
    "ier_baremo": {"Bajo": 79, "Medio": 218, "Alto": 87},
}


def theoretical_baremo(total: pd.Series) -> dict[str, int]:
    levels = pd.cut(
        total,
        bins=[-math.inf, 28, 44, math.inf],
        labels=["Bajo", "Medio", "Alto"],
        right=True,
        include_lowest=True,
    )

    if levels.isna().any():
        raise AssertionError("Hay participantes sin clasificar.")

    result = {
        level: int((levels == level).sum())
        for level in ["Bajo", "Medio", "Alto"]
    }

    if sum(result.values()) != len(total):
        raise AssertionError("Las frecuencias no suman N.")

    return result


if not DATASET.exists():
    raise FileNotFoundError(DATASET)

df = pd.read_excel(DATASET, sheet_name=0)

missing = sorted(set(PGV_ITEMS + IER_ITEMS) - set(df.columns))
if missing:
    raise AssertionError(f"Columnas faltantes: {missing}")

pgv_total = df[PGV_ITEMS].sum(axis=1, min_count=len(PGV_ITEMS))
ier_total = df[IER_ITEMS].sum(axis=1, min_count=len(IER_ITEMS))

valid = pd.concat(
    [pgv_total.rename("PGV"), ier_total.rename("IER")],
    axis=1,
).dropna()

pearson = stats.pearsonr(valid["PGV"], valid["IER"])
spearman = stats.spearmanr(valid["PGV"], valid["IER"])
kendall = stats.kendalltau(valid["PGV"], valid["IER"])

pgv_baremo = theoretical_baremo(valid["PGV"])
ier_baremo = theoretical_baremo(valid["IER"])

assert len(valid) == EXPECTED["n"], (len(valid), EXPECTED["n"])
assert abs(float(spearman.statistic) - EXPECTED["spearman_rho"]) < 1e-6
assert pgv_baremo == EXPECTED["pgv_baremo"], pgv_baremo
assert ier_baremo == EXPECTED["ier_baremo"], ier_baremo

result = {
    "dataset": str(DATASET.relative_to(ROOT)),
    "sha1": hashlib.sha1(DATASET.read_bytes()).hexdigest(),
    "n": len(valid),
    "pearson": {
        "r": float(pearson.statistic),
        "p_value": float(pearson.pvalue),
    },
    "spearman": {
        "rho": float(spearman.statistic),
        "p_value": float(spearman.pvalue),
    },
    "kendall": {
        "tau_b": float(kendall.statistic),
        "p_value": float(kendall.pvalue),
    },
    "baremos_teoricos": {
        "criterio_total": {
            "Bajo": "12–28",
            "Medio": "29–44",
            "Alto": "45–60",
        },
        "PGV": pgv_baremo,
        "IER": ier_baremo,
    },
    "status": "PASS",
}

OUTPUT.write_text(
    json.dumps(result, indent=2, ensure_ascii=False),
    encoding="utf-8",
)

print(json.dumps(result, indent=2, ensure_ascii=False))
